### PR TITLE
librados: use cursor for nobjects listing

### PR DESCRIPTION
**Pull request by yehudasa**
This includes also the changes that are in #13149, and replaces #12399. Unlike the older PR that it replaces, it leverages the ObjectCursor type that has been added for a new pool listing api (aka v3).

### CODE REVIEW

File: src/include/rados/librados.hpp [hunk 1]
```diff
@@ -72,6 +72,33 @@ namespace librados
...
+    friend std::ostream& operator<<(std::ostream& os, const librados::ObjectCursor& oc);
+
+    std::string to_str() const;
+    bool from_str(const std::string& s);
+
+    protected:
+    rados_object_list_cursor c_cursor;
+  };
+  CEPH_RADOS_API std::ostream& operator<<(std::ostream& os, const rados_object_list_cursor& c);
+  CEPH_RADOS_API std::ostream& operator<<(std::ostream& os, const librados::ObjectCursor& oc);
```
Comment (liewegas) [inline src/include/rados/librados.hpp]: Hmm, we don't do this with other parts of the librados.hpp...  doesn't hurt I guess, but is it needed?

File: src/include/rados/librados.hpp [hunk 2]
```diff
@@ -72,6 +72,33 @@ namespace librados
...
   CEPH_RADOS_API std::ostream& operator<<(std::ostream& out, const librados::ListObject& lop);
 
+  class CEPH_RADOS_API NObjectIterator;
+
+  class CEPH_RADOS_API ObjectCursor
+  {
+    public:
+    ObjectCursor();
+    ObjectCursor(const ObjectCursor &rhs);
+    ObjectCursor(rados_object_list_cursor c) : c_cursor(c) {}
```
Comment (liewegas) [inline src/include/rados/librados.hpp]: I'd add 'explicit' here so users don't accidentally do the conversion

File: src/osdc/Objecter.cc [hunk 3]
```diff
@@ -3372,6 +3372,33 @@ uint32_t Objecter::list_nobjects_seek(NListContext *list_context,
   return pos;
 }
 
+uint32_t Objecter::list_nobjects_seek(NListContext *list_context,
+				      const hobject_t *cursor)
```
Comment (liewegas) [inline src/osdc/Objecter.cc]: nit: const hobject_t& cursor here?

**Comment by liewegas** [comment]
a few nits aside this looks good to me!

File: src/include/rados/librados.hpp (hunk 1 above)
Comment (yehudasa) [inline src/include/rados/librados.hpp]: This is a copy paste of:
`CEPH_RADOS_API std::ostream& operator<<(std::ostream& out, const librados::ListObject& lop);`. Not sure whether it's needed or not.

**Comment by liewegas** [comment]
@yehudasa needs rebase

**Comment by yehudasa** [comment]
jenkins retest this please

**Comment by yehudasa** [comment]
jenkins retest this please

**Pull request merged**